File: lib/ts/ProjectLoader.ts
import { TraceProject } from '../../models/ts/Tracer_pb';

export abstract class ProjectLoader {
    constructor() {
    }

    public async LoadProject(id: string): Promise<TraceProject> {
        const traceProject: TraceProject = TraceProject.deserializeBinary(await this.GetProjectStream(id));

        return traceProject;
    }

    protected abstract async GetProjectStream(id: string): Promise<Uint8Array>;
}

File: lib/cs/ProjectLoader.cs
using System;
using System.IO;
using System.Threading.Tasks;
using Tracer;

public abstract class IProjectReader
{
    IProjectReader()
    {
    }

    public async Task<TraceProject> LoadProject(Guid id)
    {
        using (var stream = await this.GetProjectStream(id))
        {
            TraceProject traceProject = TraceProject.Parser.ParseFrom(stream);
            return traceProject;
        }
    }

    protected abstract Task<Stream> GetProjectStream(Guid id);
}
